bluesky/simulators.py

```diff
@@ -114,3 +114,276 @@ def check_limits(plan):
+            #the number of 'triggers' called for later use in the time estimate.
+            if msg.command == 'set': 
+                val_dict['set'][msg.obj.name] = msg.args[0]
+
+            elif msg.command == 'trigger': 
+                if msg.obj.name in list( val_dict['trigger'].keys() ):
+                    val_dict['trigger'][msg.obj.name] += 1
+
+                else:
+                    val_dict['trigger'][msg.obj.name] = 1
+
+            #this section estimates the time fro each command.
+            msg = next(plan)
+            object_est_time, val_dict = obj_est_time(msg, val_dict)
+            out_est_time = combine_est_time(out_est_time, object_est_time, method = 'max')
+
+        return out_est_time, val_dict
+
+
+    def run_est_time(msg, val_dict):
+        """
+        Returns the est_time/std_dev pair for a run.
+        This function returns an est_time/std_dev pair, for a group, where a group is defined as a 
+        series of messages which are run simultaneously and is ended on a wait messages. It 
+        assumes that it is called from inside plan_est_time, once an 'open_run' has been detected. 
+        The routine also returns an updated version of val_dict including any changed values 
+        from the run.
+
+        Parameters
+        ----------
+        msg : message.
+            The first message of the group that was detected in plan_est_time.
+        val_dict : dict.
+            A dictionary containing information on values updated during the plan. It has key:arg 
+            pairs with keys relating to message components where each arg is a dictionary 
+            containing devices, (detectors, axes) that have been updated, as keywords and a value 
+            indicating the update status. For 'set' the update status is the latest 'position' 
+            that it has reached and for 'trigger' the update status is the number of times since 
+            the beginning of the scan or since an 'unstage' event that the device has been 
+            triggered.
+        
+        Return Parameters
+        -----------------
+        out_est_time : list.
+            The combined est_time/std_dev pair.
+        val_dict : dict.
+            The updated version of val_dict.
+        """
+
+        out_est_time = [0,0]
+        
+        while msg.command not in _run_end_cmds:
+            msg = next(plan)
+            if msg.command in (_run_start_cmds + _flyer_start_cmds):
+                grp_est_time, val_dict = group_est_time(msg, val_dict)
+                out_est_time = combine_est_time(out_est_time, grp_est_time)
+
+            else:
+                object_est_time, val_dict = obj_est_time(msg, val_dict)
+                out_est_time = combine_est_time(out_est_time, object_est_time)
+        return out_est_time, val_dict
+
+
+    #Define some variables used in the following.
+    val_dict = {'set':{}, 'trigger':{} } #this holds information on the updated values.
```

Comment:
Now for a signature naming-related nitpick! :-D

I don't find `val_dict` particularly descriptive. I usually see variables with the kind role with `state` in the name, which also hints that this is is something we are going to mutating a lot. Maybe just name it `state`?

Reply:
`state` means *so* many different things and is likely to already have implementation in cases where time estimates would be a valuable add-on.  At a loss how to make this name more specific (and less likely to conflict)-- ideas?

Reply:
I too am not so happy with val_dict, I will think about what else it can be but don't like state for the same reasons as @prjemian

Reply:
`msg_info` ?  Seems concise, descriptive for its context, and not likely in use

Reply:
This is pretty well scoped (it is not actually publicly exposed at all right now!) so I am not worried about collisions with other uses of `state`.  I like `state` quite a bit here, but I think that this is a candidate for a class (like the RE) to handle this via the visitor pattern (ex https://en.wikipedia.org/wiki/Visitor_pattern ).

Reply:
ok with naming.  Agree there is more complexity here; a class is warranted.

Reply:
I quite like msg_info, but maybe plan_info is better as it is keeping track of what items have been 'modified' in the plan up to this point. As I mentioned above I am not exactly sure how this would work as a class, but will think about it.

Reply:
To try to clarify the use of val_dict, essentially to know how long it takes to 'move' a device it is necessary to know the starting point. As this can be modified by an earlier message in the plan when stepping through messages without performing the action I need to keep track of any previous changes to this that would have been caused by the plan itself. There are also other changes (settle_time etc) that may be 'changed' during a plan, and can change the time estimate, that need to be tracked.

Reply:
Complex numbers are probably the best example I can think of.  Special math operations for them.  This `msg_info` (or is it now `plan_history`) class could combine the special operations with the historical info such as a database.  Not sure I well understand what is required.  Assume the doc strings will describe in some detail.

Reply:
I thin the docstrings already describe pretty well, if not let me know and I will update them. Essentially, as an example, think of an 'area detector' detector. To a first approximation the est_time for a triggering can be written as est_time = settle_time + num_images * acquire_time.

In a plan we could easily change the acquire_time part of the way through the plan. If this happens then we would need this 'updated' value in order to estimate the time for any following triggers of this detector in the plan. As the generator function does not actually perform this change, it needs to be 'tracked' which is what I use val_dict(name change pending) for.

Similar things occur for other actions (set for example by definition changes values). Hopefully this is a little clearer. I think I have a better way to do this using classes, I will write it up and see how it goes.